File: solutions/1760. Minimum Limit of Balls in a Bag/1760.py
# Time complexity: O(nlog(max(nums)))
# Space complexity: O(1)
class Solution:
  def minimumSize(self, nums: List[int], maxOperations: int) -> int:
    def numOperations(m: int) -> int:
      return sum((num - 1) // m for num in nums) <= maxOperations
    return bisect.bisect_left(range(1, max(nums)), True,
                              key=lambda m: numOperations(m)) + 1

File: solutions/1760. Minimum Limit of Balls in a Bag/1760.rs
// Time complexity: O(nlog(max(nums)))
// Space complexity: O(1)
impl Solution {
    pub fn minimum_size(nums: Vec<i32>, max_operations: i32) -> i32 {
        fn num_operations(nums: &Vec<i32>, m: i32) -> i32 {
            nums.iter().map(|&num| (num - 1) / m).sum()
        }

        let mut l = 1;
        let mut r = *nums.iter().max().unwrap();

        while l < r {
            let m = l + (r - l) / 2;
            if num_operations(&nums, m) <= max_operations {
                r = m;
            } else {
                l = m + 1;
            }
        }

        l
    }
}
